tools/wasm-sourcemap.py

```diff
@@ -264,6 +264,14 @@ def build_sourcemap(entries, code_section_offset, prefixes, collect_sources):
       column = 1
     address = entry['address'] + code_section_offset
     file_name = entry['file']
+    # prefer to emit a relative path to the base path, which is where the
```

Comment:
I'm not sure if we should do anything special here... since we're talking about edge cases anyway, consider one where a static server that is hosting `/` directory under, say, `http://somedomain/static/`.

Then, let's say Wasm is under `/dist/temp.wasm` on the disk (corresponding to URL: `http://somedomain/static/dist/temp.wasm` on server), and source is under `/src/temp.c` (URL: `http://somedomain/static/src/temp.c`).

If you use relative paths, then source would be encoded as `../src/temp.c` and resolve correctly in URL context, but if you keep it as-is, then it would be encoded as `/src/temp.c` and resolve to `http://somedomain/src/temp.c` in browser, which is invalid.

---

TL;DR - let's just do the simple thing and use relative paths unconditionally, which would happen to also cover edge cases more correctly.

Reply:
That would break the current `other.test_wasm_sourcemap` test, it gets this as the source file location:
```
"wasm-src://../../emscripten/tests/other/wasm_sourcemap/no_main.c"
```
Is that ok?

Reply:
(if that's not ok, what would the proper output there be, do you think?)

Reply:
I see, interesting. Where is `wasm-src://` added? We probably should skip resolving relative paths altogether when that option is enabled.

Reply:
It's added in the test, using the `--prefix` flag to the source map tool (which I don't understand well).

Maybe we can remove that option? If you and @yurydelendik agree that should be fine as I don't see other users in the codebase.

Reply:
The DWARF information must contain absolute path to source file in .debug_line and .debug_info. If for some reason we don't have that, it needs to be fixed. The `--prefix` was designed to perform the basepath logic, e.g. "http://somedomain/static/dist/temp.wasm + http://somedomain/src/temp.c" with use of `--prefix /Users/me/example/=../src/` shall produce the desired result. I wonder if wasm-sourcemap.py not reading/building entire DWARF file path.

Reply:
@yurydelendik isn't adding a prefix what `--source-map-base` is for? (which does not use `--prefix` under the hood) Or is that for something else?

Reply:
`--source-map-base` is instruction for tools to add proper map location in sourceMappingURL, this is optional. (As defined in https://github.com/WebAssembly/design/pull/1051) Devtools use `sourceMappingURL` as base URL for all source names.

`--prefix`, on server/build side, helps to strip useless for devtools absolute part. Notice that sources paths needs to point to the valid internet resource only `sourcesContent` is not present.

Reply:
It looks like nothing uses `--prefix`, though - I think we are avoiding the problem you describe with the other tools here already? Or are you aware of out-of-tree users of wasm-sourcemap.py? If not, would you object to me removing that flag?

Reply:
`--prefix` is designed to match gcc's `-fdebug-prefix-map=` . I don't have objection, except convenience, since clang or rustc can do it. Will `emcc` passthru this option?

> are you aware of out-of-tree users of wasm-sourcemap.py

wasm-sourcemap.py is designed and is used in only for emscripten (and my examples).

Reply:
Interesting, I wasn't aware of `-fdebug-prefix-map`. emcc does not currently pass that through. Perhaps it should?

Also, getting back to the more immediate issue here: It seems like the fundamental issue is that when supplying `--prefix`, we should not automatically relativize a path. That is what this PR does currently, but @RReverser wrote this comment suggesting we always relativize. To me it sounds like the PR should stand as it is - @RReverser what do you think?

Reply:
1. No, if we keep it, I also suggested that we shouldn't relativize when `--prefix` is passed:
> We probably should skip resolving relative paths altogether when that option is enabled.
2. However, it's not what the PR does at the moment - right now it checks whether paths don't have a shared prefix. This check is incorrect and breaks cases like described in comment (1) of this thread. Instead, it should skip relativising specifically when `--prefix` is set.
3. Finally, maybe I misunderstood, but from the exchange above it sounds like we can remove `--prefix` instead?

Reply:
> emcc does not currently pass that through. Perhaps it should?

Yes, it should. After removal `--prefix` it will be the only way to properly re-point source to right location.

Here is a use case: there are several libraries that are checked out locally on a build machine. The desired results is that source with point to proper internet resources. The only way to do provide right source location is to use `-fdebug-prefix-map`, e.g. `-fdebug-prefix-map=/Users/build/lib1/=http://example.org/lib1/v1/ --fdebug-prefix-map=/Users/build/lib2/=http://example.org/lib2/v9/`.

Reply:
@yurydelendik Doesn't `-fdebug-prefix-map` in Clang change DWARF paths in-place to the remapped locations? If it does, then what is the role of custom `--prefix` handling?

Reply:
> what is the role of custom --prefix handling

To do it after the fact of creating a wasm binary. (`-fdebug-prefix-map` did not work in rustc at the moment of writing wasm-sourcemap.py)

Reply:
Sure, but now that both compilers support native remapping, it seems that we can just skip it here?

Reply:
the `--base_path` logic is a really narrow case of the `--prefix`. The former assumes that sources folders will have the same directory structure when published and calculates replacement part automatically, the latter provides more flexibility.  FWIW they can exist together (by adding `else:` for `prefixes.sources.resolve` line). Sorry, I cannot provide more definite "yes, let's remove it", since I accustomed to the `--prefix` way. :)

Reply:
Thanks @RReverser , you are right, I had that mixed up! Should be correct now - use the prefixes if provided, otherwise use a proper relative path.